### PR TITLE
DOCS: Adding DOCA_UROM readme

### PR DESCRIPTION
**Pull request by janjust**
## What
Add a DOCA_UROM and DOCA_UROM plugin readme into the contrib folder of DOCA_UROM plugin

## Why ?
To better explain the plugin and the how to build/run

## How ?
it's a markdown file with instructions and scripts of example runs.

### CODE REVIEW

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# Plugin Description
+The CL_DOCA_UROM plugin enables collective offloads via DOCA and CL_DOCA_UROM feature. The plugin runs on the DPU and is initialized via the command channel from CL_DOCA_UROM. It leverages UCC collectives optimized for the DPU and facilitates efficient computation and communication overlap. While the plugin supports all UCC collectives via a copy-in/out method, the optimized algorithms are Allreduce and Alltoall/v (OpenSHMEM).
+
+## Overall Design and System Architecture
```
Comment (manjugv) [inline contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md:4]: @janjust  Let's change this heading to components and dependencies?

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+2. **DPU:**
+    - Runs the DOCA services.
+    - Leverages UCC for data handling.
+    - Communicates with the host node using the DOCA interface.
+
+### Interaction
+- The host node communicates with the DPU via the DOCA interface.
+- Data and commands are sent from the host to the DPU, processed by UCC, and results are returned.
+
+### Technology Stack
```
Comment (manjugv) [inline contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md:19]: Dependencies?

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
+# Plugin Description
```
Comment (manjugv) [inline contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md:1]: Plugin Description -> "How to enable and run collectives on NVIDIA DPUs?"

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md (hunk 1 above)
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: done

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md (hunk 2 above)
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: done

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md (hunk 3 above)
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: done

**Review by manjugv — APPROVED** [review]

**Comment by manjugv** [comment]
@samnordmann  can you review?

**Review by samnordmann — APPROVED** [review]
LGTM!
I left a bunch of minor remarks, feel free to discard those you think are not relevant

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    ./configure --prefix=<path> --with-ucx=<path to ucx>  --with-doca_urom=<path to doca>
+    make -j install
+    ```
+
+### DPU
+1. **Build Instructions:**
+    - Checkout the same UCC branch as on the host.
+    - Use the same configure line as the host.
+
+### Generate the following files:
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: fixed, I think, I added that those file are to be generated in $TOP

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+### 2. Run on Host:
+- **Script:** `run_doca_urom_cl.sh`
+    ```bash
+    #!/bin/bash
+
+    PPN=$1
+
+    if [ -z "$PPN" ]; then
+        echo dollar 1=PPN not set, assuming 1 PPN
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: fixed

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+## Run Instructions
+
+### 1. Run on DPU:
+- **Script:** `run_doca_uromd.sh`
+    ```bash
+    #!/bin/bash
+
+    TOP=<working dir>
+
+    ompi=$TOP/<ompi dir>
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: agreed, changed the naming convention

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    options="$options -x DOCA_UROM_SERVICE_FILE=$TOP/servicefile"
+    options="$options -x LD_LIBRARY_PATH=$LD_LIBRARY_PATH"
+
+    # Comma separated list of envs
+    options="$options -x UCC_CL_DOCA_UROM_PLUGIN_ENVS=LD_LIBRARY_PATH=$TOP/arm/build-arm/ompi/lib:$TOP/arm/build-arm/ucx/lib:$TOP/arm/build-arm/ucc/lib:$TOP/doca_urom_ucc/install/dpu/lib64,UCX_LOG_LEVEL=ERROR"
+
+    options="$options --mca coll_ucc_enable 1 --mca coll_ucc_priority 100"
+    options="$options -x UCX_TLS=rc_x,tcp"
+
+    validation="--validation"
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: I left this one, but I added that it's an optional argument to osu bench to run data validation

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    msg_low=1024
+    msg_high=$((1024*1024))
+
+    for coll in allreduce
+    do
+        # urom cl
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: removed

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    PPN=$1
+
+    if [ -z "$PPN" ]; then
+        echo dollar 1=PPN not set, assuming 1 PPN
+        PPN=1
+    fi
+
+    TOP=<path to build dir>
+
+    NODES=$(cat $TOP/hostfile | grep -v '#' | wc -l)
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: changed

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    TOP=<working dir>
+
+    ompi=$TOP/<ompi dir>
+    dpufile=$TOP/dpufile
+
+    NODES=$(cat $dpufile | grep -v '#' | wc -l)
+
+    # Set plugin path
+    export UROM_PLUGIN_PATH=<path to ucc doca_plugins>
+    export PATH=$ompi/bin:$PATH
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: technically no, once exported to path it makes no difference, but it was a habit of mine to set PATH and LD_LIBRARY_PATH to the working ompi and then still use full path for clarity - but changed it anyways to not confuse users

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+- **APIs:** UCC collective operations with DOCA_UROM extensions.
+
+## Build and Deployment
+
+For DOCA instructions please see: [DOCA - Get Started | NVIDIA Developer](https://docs.nvidia.com/doca/sdk/nvidia+doca+overview/index.html)
+
+### Host Node
+1. **Build Instructions:**
+    ```sh
+    Checkout UCC
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md:35]: nice catch - looks like it was a typo or a leftover from the actual build script copied into the markdown convertor

File: contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    options="$options -x UCX_TLS=rc_x,tcp"
+
+    validation="--validation"
+    msg_low=1024
+    msg_high=$((1024*1024))
+
+    for coll in allreduce
+    do
+        # urom cl
+        $OMPI/bin/mpirun -np $((NODES * PPN)) --map-by node -hostfile $TOP/hostfile $options --mca coll_ucc_cls doca_urom,basic --tag-output $OMB/osu_$coll $validation "-m $msg_low:$msg_high -M $((1*1024*1024*1024))" -i 10 -x 5
```
[the remark this answers is not in the document]
Comment (janjust) [reply contrib/doca_urom_ucc_plugin/DOCA_UROM_README.md]: removed and only left -m min:high, since that's enough to run this, I don't recall what -M was

**Comment by janjust** [comment]
thanks for the comments @samnordmann , I fixed, addressed all of them, useful stuff

**Pull request merged**